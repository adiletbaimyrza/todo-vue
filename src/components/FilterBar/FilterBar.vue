<template>
  <div :class="$style['container']" :data-testid="DataTestIds.FILTER_BAR">
    <p
      :class="$style['container__count']"
      :data-testid="DataTestIds.ACTIVE_COUNT"
    >
      {{ activeCount }} items left
    </p>
    <ul
      :class="$style['container__filters']"
      :data-testid="DataTestIds.FILTERS"
    >
      <li>
        <button
          @click="setFilter(TodoFilter.ALL)"
          type="button"
          :class="{
            [$style['container__filters--active']]:
              currentFilter === TodoFilter.ALL,
          }"
          :data-testid="DataTestIds.FILTER_ALL"
        >
          all
        </button>
      </li>
      <li>
        <button
          @click="setFilter(TodoFilter.ACTIVE)"
          type="button"
          :class="{
            [$style['container__filters--active']]:
              currentFilter === TodoFilter.ACTIVE,
          }"
          :data-testid="DataTestIds.FILTER_ACTIVE"
        >
          active
        </button>
      </li>
      <li>
        <button
          @click="setFilter(TodoFilter.COMPLETED)"
          type="button"
          :class="{
            [$style['container__filters--active']]:
              currentFilter === TodoFilter.COMPLETED,
          }"
          :data-testid="DataTestIds.FILTER_COMPLETED"
        >
          completed
        </button>
      </li>
    </ul>
    <button
      @click="deleteCompleted"
      type="button"
      :class="$style['container__clear']"
      :data-testid="DataTestIds.CLEAR_COMPLETED"
    >
      clear completed
    </button>
  </div>
</template>

<script>
import { mapState, mapGetters, mapMutations } from 'vuex'
import { TodoFilter, DataTestIds } from '@/constants'

export default {
  name: 'FilterBar',
  computed: {
    ...mapState(['currentFilter']),
    ...mapGetters(['activeCount']),
  },
  methods: {
    ...mapMutations(['setFilter', 'deleteCompleted']),
  },
  data() {
    return {
      TodoFilter,
      DataTestIds,
    }
  },
}
</script>

<style lang="scss" module>
.container {
  display: flex;
  padding: 1em 1.5em;
  align-items: center;
  width: 600px;
  background-color: var(--card-background-color);

  &__count {
    font-size: 0.9em;
    color: #b1b1b1;

    &:hover {
      cursor: default;
    }
  }

  &__filters {
    flex-grow: 1;
    display: flex;
    justify-content: center;

    li button {
      color: #b1b1b1;
      font-family: Inter, sans-serif;
      font-weight: 500;
      padding: 0 1em;
      text-transform: capitalize;

      &:hover {
        color: var(--filter-text--hover);
      }
    }

    &--active {
      color: #1969ff !important;
    }
  }

  &__clear {
    font-family: Inter, sans-serif;
    text-transform: capitalize;
    color: #b1b1b1;

    &:hover {
      color: var(--filter-text--hover);
    }
  }
}
</style>
