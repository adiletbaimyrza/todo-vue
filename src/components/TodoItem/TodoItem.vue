<template>
  <div @click="checkTodo" :class="$style['item']">
    <button
      @click.stop="checkTodo"
      :class="$style['item__button']"
      type="button"
    >
      <CoreIcon
        :src="
          isCompleted
            ? require('../../assets/circle-check.svg')
            : require('../../assets/circle.svg')
        "
      />
    </button>
    <p
      :class="[
        $style['item__content'],
        { [$style['item__content--completed']]: isCompleted },
      ]"
    >
      {{ content }}
    </p>
    <button
      @click.stop="deleteTodo"
      :class="[$style['item__button'], $style['item__button--delete']]"
      type="button"
    >
      <CoreIcon :src="require('../../assets/xmark.svg')" />
    </button>
  </div>
</template>

<script>
import CoreIcon from '../CoreIcon/CoreIcon.vue'

export default {
  name: 'TodoItem',
  components: {
    CoreIcon,
  },
  props: {
    id: {
      type: Number,
      required: true,
    },
    content: {
      type: String,
      required: true,
    },
    isCompleted: {
      type: Boolean,
      default: false,
    },
  },
  methods: {
    checkTodo() {
      this.$store.commit('check', this.id)
    },
    deleteTodo() {
      this.$store.commit('delete', this.id)
    },
  },
}
</script>

<style lang="scss" module>
.item {
  display: flex;
  width: 600px;
  padding: 1em 1.5em;
  align-items: center;
  color: var(--font-color-regular);
  background-color: var(--card-background-color);
  border-bottom: 1px solid var(--card-border-color);

  &:hover {
    background-color: var(--card-background-color--hover);
    cursor: pointer;

    .item__button--delete {
      visibility: visible;
    }
  }

  &__content {
    flex-grow: 1;
    text-align: left;
    padding: 0 1em;

    &--completed {
      color: #b1b1b1;
      text-decoration: line-through;
    }
  }

  &__icon {
    height: 2em;
    width: 2em;
  }

  &__button {
    &--delete {
      visibility: hidden;
    }
  }
}
</style>
