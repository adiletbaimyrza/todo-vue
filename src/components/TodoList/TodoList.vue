<template>
  <div :class="$style['list']">
    <TodoItem
      v-for="item in filteredTodos"
      :key="item.id"
      :id="item.id"
      :content="item.content"
      :isCompleted="item.isCompleted"
    />
    <FilterBar />
  </div>
</template>

<script>
import { mapGetters } from 'vuex'
import FilterBar from '../FilterBar/FilterBar.vue'
import TodoItem from '../TodoItem/TodoItem.vue'

export default {
  name: 'TodoList',
  components: {
    TodoItem,
    FilterBar,
  },
  computed: {
    ...mapGetters(['filteredTodos']),
  },
}
</script>

<style lang="scss" module>
.list {
  display: flex;
  flex-direction: column;
  box-shadow: rgb(100 100 111 / 20%) 0 7px 29px 0;
  border-radius: 0.4em;

  > :first-child {
    border-top-left-radius: 0.4em;
    border-top-right-radius: 0.4em;
  }

  > :last-child {
    border-bottom-left-radius: 0.4em;
    border-bottom-right-radius: 0.4em;
    border-bottom: none;
  }
}
</style>
