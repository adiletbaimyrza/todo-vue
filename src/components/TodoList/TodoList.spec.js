import { shallowMount } from '@vue/test-utils'
import { createStore } from 'vuex'
import TodoList from './TodoList.vue'
import TodoItem from '../TodoItem/TodoItem.vue'
import FilterBar from '../FilterBar/FilterBar.vue'

/* global describe beforeEach expect it */
describe('TodoList.vue', () => {
  let store
  let state
  let getters
  let wrapper

  beforeEach(() => {
    state = {
      todos: [
        { id: 1, content: 'Task 1', isCompleted: false },
        { id: 2, content: 'Task 2', isCompleted: true },
      ],
    }

    getters = {
      filteredTodos: () => state.todos,
    }

    store = createStore({
      state,
      getters,
    })

    wrapper = shallowMount(TodoList, {
      global: {
        plugins: [store],
      },
    })
  })

  it('renders correctly with initial state', () => {
    expect(wrapper.exists()).toBe(true)
    expect(wrapper.element).toMatchSnapshot()
  })

  it('renders the correct number of TodoItem components', () => {
    const todoItems = wrapper.findAllComponents(TodoItem)
    expect(todoItems.length).toBe(state.todos.length)
  })

  it('passes the correct props to TodoItem components', () => {
    const todoItems = wrapper.findAllComponents(TodoItem)
    todoItems.forEach((item, index) => {
      expect(item.props().id).toBe(state.todos[index].id)
      expect(item.props().content).toBe(state.todos[index].content)
      expect(item.props().isCompleted).toBe(state.todos[index].isCompleted)
    })
  })

  it('renders the FilterBar component', () => {
    const filterBar = wrapper.findComponent(FilterBar)
    expect(filterBar.exists()).toBe(true)
  })
})
