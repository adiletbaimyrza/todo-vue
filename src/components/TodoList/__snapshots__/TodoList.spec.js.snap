// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList.vue renders correctly with initial state 1`] = `
<div
  class="list"
>
  
  <todo-item-stub
    content="Task 1"
    id="1"
    iscompleted="false"
  />
  <todo-item-stub
    content="Task 2"
    id="2"
    iscompleted="true"
  />
  
  <filter-bar-stub />
</div>
`;
